apps/builder/app/builder/builder.tsx

```diff
@@ -208,6 +223,108 @@ const NavigatorPanel = ({
   );
 };
 
+const revealAnimation = ({
+  show,
+  backgroundColor,
+}: {
+  show: boolean;
+  backgroundColor: string;
+}): CSS => ({
+  position: "relative",
+  "> ::after": {
+    content: "",
+    position: "absolute",
+    inset: 0,
+    zIndex: 1,
+    transitionDuration: "300ms",
+    pointerEvents: "none",
+    transitionProperty: "opacity",
+    backgroundColor,
+    opacity: show ? 0 : 1,
+  },
+});
+
+const useLoadingState = () => {
+  type State =
+    | "dataLoadingState"
+    | "selectedInstanceRenderState"
+    | "canvasIframeState";
+  const [states, setStates] = useState<Map<State, "initial" | "ready">>(
+    new Map([
+      ["dataLoadingState", "initial"],
+      ["selectedInstanceRenderState", "initial"],
+      ["canvasIframeState", "initial"],
+    ])
+  );
+
+  const dataLoadingState = useStore($dataLoadingState);
+  const selectedInstanceRenderState = useStore($selectedInstanceRenderState);
+  const canvasIframeState = useStore($canvasIframeState);
+  const update = (name: State, current: string, expected: string) => {
+    if (current === expected && states.get(name) !== "ready") {
+      states.set(name, "ready");
+      setStates(new Map(states));
+    }
+  };
+
+  // Each state can have a different readiness value.
+  update("dataLoadingState", dataLoadingState, "loaded");
+  update("selectedInstanceRenderState", selectedInstanceRenderState, "mounted");
+  update("canvasIframeState", canvasIframeState, "ready");
+
+  const readyCount = Array.from(states.values()).filter(
+    (state) => state === "ready"
+  ).length;
+  const state = readyCount === states.size ? "ready" : "loading";
+  const progress = Math.round((readyCount / states.size) * 100);
+
+  return { state, progress, states };
+};
+
+const ProgressIndicator = ({ value }: { value: number }) => {
```

Comment:
This is something what starts from `value` so value is `initialValue` here

Reply:
I have 2 values, real value, which is value and a fakeValue. Its not just initial, because its updating, semantically in useState initial value is not updating